incubator/etcd/templates/configmap.yaml

```diff
@@ -0,0 +1,176 @@
+      etcdctl member list | grep http://${HOSTNAME}.${SET_NAME}.${POD_NAMESPACE}:2379 | cut -d':' -f1 | cut -d'[' -f1 > /etcd/member_id
+      exit 0
+    }
+
+    # endpoint of the first etcd member
+    ep() {
+      echo "http://${SET_NAME}-0.${SET_NAME}.${POD_NAMESPACE}:2379"
+    }
+
+    # wait for the given hostname to become available to the network
+    await_host() {
+      echo "Waiting for [${1}] to come up "
+      while true; do
+        echo -n '.'
+        ping -W 1 -c 1 ${1} > /dev/null && break
+        sleep 1s
+      done
+      echo -n " done"
+    }
+
+    # add a new member to the etcd cluster
+    add_member() {
+      EPS=$(ep)
+      MEMBER_LIST=$(etcdctl --endpoints ${EPS} member list)
+
+      if echo "${MEMBER_LIST}" | grep ${HOSTNAME}; then
+        echo "Removing old member [${HOSTNAME}] from etcd cluster"
+        MEMBER_ID=$(echo "$MEMBER_LIST" | grep ${HOSTNAME} | cut -d':' -f1 | cut -d'[' -f1)
+        etcdctl --endpoints ${EPS} member remove ${MEMBER_ID}
+      fi
+
+      if [ -e $ETCD_DATA_DIR ]; then
+        echo "Removing old data dir [${ETCD_DATA_DIR}] "
+        rm -Rf $ETCD_DATA_DIR
+      fi
+
+      echo "Adding [${HOSTNAME}] as a new member"
+      STR_ENV="ETCD_"
+      ETCD_PREFIX=$(etcdctl --endpoints ${EPS} member add ${HOSTNAME} ${PROTO}://${HOSTNAME}.${SET_NAME}.${POD_NAMESPACE}:2380 | \
+      {
+        while read i
+          do
+            if test "${i#*$STR_ENV}" != "$i"; then
+              ETCD_PREFIX="$ETCD_PREFIX $i"
+            fi
+          done
+        echo "export $ETCD_PREFIX" | sed 's/"//g'
+      })
+      eval \${ETCD_PREFIX}
+
+      collect_member &
+      echo "Start etcd daemon"
+      exec etcd \
+        --listen-peer-urls ${PROTO}://${IP}:2380 \
+        --initial-advertise-peer-urls ${PROTO}://${HOSTNAME}.${SET_NAME}.${POD_NAMESPACE}:2380 \
+        --listen-client-urls http://${IP}:2379,http://127.0.0.1:2379 \
+        {{- if .Values.etcd.peerTLS }}
+        --peer-auto-tls \
+        {{- end }}
+        --advertise-client-urls http://${HOSTNAME}.${SET_NAME}.${POD_NAMESPACE}:2379
+    }
+
+    # <release-name>-etcd-<set-id>
+    SET_ID=${HOSTNAME##*-}
+
+    # are we rejoining after a failure?
+    if [ -e /etcd/member_id ]; then
+        echo "Found old member ID, attempting to rejoin the etcd cluster"
+
+        MEMBER_ID=$(cat /etcd/member_id)
+
+        # if we are the first member we can't resolve over the default endpoint
+        # attempt to resolve over the defined service instead
+        if [ "${SET_ID}" -eq 0 ]; then
+          EPS="http://${SET_NAME}.${POD_NAMESPACE}:2379"
+        else
+          EPS=$(ep)
+        fi
+
+        await_host ${HOSTNAME}.${SET_NAME}.${POD_NAMESPACE}
+
+        # attempt to resolve a member list from the selected endpoint
+        # if this fails we assume that the previous cluster is gone
+        set +e
+        MEMBER_LIST=$(etcdctl --endpoints ${EPS} member list)
+        if [ "$?" -ne 0 ]; then
```

Comment:
This is not introduced by me but dangerous. If your whole k8s cluster goes down for whatever reason this will wipe your `etcd` storage as soon as your cluster starts recovering.

Reply:
Ouch, we shouldn't do that. Why does it wipe the storage and not just keep retrying to create a cluster from the pieces it has and/or can find? I'm not very proficient with etcd, mind you. But I'm sure we don't want our cluster wiped in any case if we did not throw away the PVs ourselves...

Reply:
My thoughts exactly...

Etcd isn't able to recover from a complete failure without some additional procedures which are described here: https://coreos.com/etcd/docs/latest/op-guide/recovery.html

Reply:
Your effort is very much appreciated!

Reply:
@hiddeco Are you still working on the 'disaster recovery' part? 
In case of lost quorum - do you plan to copy snapshot from one node to others?